add generated clients. modify codegen script

maleck13 opened the pull request:
**What this PR does / why we need it**:
Adds in the generated clientsets for the sample apiserver. Modifies the update-codgen script to copy over the client from the generated location. We need it in order to be able to add integrations and unit tests that make use of the clientsets and their fakes.

**Special notes for your reviewer**:
@deads2k  @p0lyn0mial  hopefully done the correct thing here. Not 100% sure on needing the copy but it seemed what was needed to get the import paths correct?

**Release note**:

```release-note
NONE
```

maleck13 committed:
Add generated clients. modify codegen script to make modification easier and to allow it to work from the root of the sample server.

fix issue where binaries not being removed. Add new client packages to lint

put new linted packages in correct location. remove flag from variables

Add all pkgs that should be linted and run sort

Add output from hack/update-bazel.sh

remove genclient=true from FischerList and regenerate client

re update bazel build files

fix missing resource method in register.go

sttts commented on staging/src/k8s.io/sample-apiserver/hack/update-codegen.sh:
```diff
@@ -67,3 +67,5 @@ ${informergen} \
   --output-package k8s.io/sample-apiserver/pkg/client/informers
   "$@"
+
+cp -R $GOPATH/src/k8s.io/sample-apiserver/pkg/client $GOPATH/src/k8s.io/kubernetes/staging/src/k8s.io/sample-apiserver/pkg/  
```
Does this mean that the upper clientgen creates the client at the wrong location?

maleck13 replied:
It does currently correct. It generates at ```$GOPATH/src/k8s.io/sample-apiserver/pkg/client``` I will look through the script commands and see can I improve this.

p0lyn0mial commented:
@maleck13 Could you rebase your branch and regenerate the `clientset` ? I have added a new type recently.

https://github.com/kubernetes/kubernetes/pull/48444

maleck13 commented:
@p0lyn0mial regenerated.
@sttts I have updated the script now so that is will generate the clients in the right location

maleck13 commented:
Apologies just noticed I have generated additional packages in the client package ```listers`` and ```listers_generated``` along with ```informers``` and ```informers_generated```

p0lyn0mial commented:
@maleck13 many thanks.

sttts commented on staging/src/k8s.io/sample-apiserver/hack/update-codegen.sh:
```diff
@@ -18,8 +18,15 @@ set -o errexit
 
-KUBE_ROOT=$(dirname "${BASH_SOURCE}")/../../../../..
-EXAMPLE_ROOT=$(dirname "${BASH_SOURCE}")/..
+KUBE_ROOT=${GOPATH}/src/k8s.io/kubernetes
```
that are a lot of variables now 👀

maleck13 replied:
Yes the intention was to make it easier for someone using the sample-apiserver as a base to modify the script and use it for their own purposes.

sttts replied:
To make this reusable and easier to read: what about turning the generator calls into one big bash functions that gets GROUP_NAME, VERSION as parameters:
```
generate_group wardle v1alpha1
```
Then only BASE_PATH and BASE_PKG would be global variables. The other *_PKG variables can become local variables in the generate_group func.

maleck13 replied:
sure that seems reasonable

sttts commented on staging/src/k8s.io/sample-apiserver/hack/update-codegen.sh:
```diff
@@ -32,38 +39,41 @@ trap cleanup EXIT
+  --listers-package ${LISTERS_PKG} \
+  --output-package ${INFORMERS_PKG} \
+  --output-base ${SERVER_BASE}
+  "$@"
```
newline missing

maleck13 commented:
@sttts I believe I have made the change you were asking for

sttts commented on staging/src/k8s.io/sample-apiserver/hack/update-codegen.sh:
```diff
@@ -30,40 +32,53 @@ function cleanup {
-  --listers-package k8s.io/sample-apiserver/pkg/client/listers \
-  --output-package k8s.io/sample-apiserver/pkg/client/informers
-  "$@"
+generate_group wardle v1alpha1
```
looks much better.

One question: if there is more than one group or version, can this script be easily adapted? Can you try to clone the existing group locally and see how to adapt the script? Would be great if this script could be used as a blue print for people building apiservers.

maleck13 replied:
Great.
On your questions. 
If there is more than one group or version, I think it would be relatively easy. But I can try this out. I am relatively new to apiservers, is there any tool for generating the scaffold of an api group and version, or is it case of copy and modify?

One your second question. I am not sure I follow. When you say "the existing group" are you referring to the repo https://github.com/kubernetes/sample-apiserver or something different?

sttts replied:
This repo is meant as a blue print. 

There is also https://github.com/kubernetes-incubator/apiserver-builder which is meant as an alternative, opinionated way to create apiservers.

Would be great to test it with multiple groups/versions. It might be that you have to pass all of them at once to clientgen because the created clients will share code. So maybe the structure of the function must be different.

maleck13 replied:
@sttts you are correct. When generating for multiple groups in its current form the script will generate only for the last group passed.
I can take a look at improving this. One question though as I believe @p0lyn0mial is wanting to use the generated clients and the proposed script will successfully generate the clients which it wouldn't before, is there enough value in the current PR to move forward with it and then follow up with improvements to the generate script? Or would you prefer we keep this open?

sttts replied:
If @p0lyn0mial is blocked by this, it's fine to merge as is and improve in a follow-up. Otherwise, it wouldn't mind to have it here.

p0lyn0mial replied:
yes, I need the `clientset` in my admission plugin. Thanks.

sttts commented:
/approve no-issue

sttts commented:
@maleck13 can you squash?

sttts commented:
/ok-to-test

maleck13 commented:
squashed

sttts commented:
/lgtm

maleck13 commented:
/retest

sttts commented:
@maleck13 please rebase. At least the unit test flake is fixed.

maleck13 commented:
rebased. FYI made minor modification to the script as I noticed the generated binaries were not being removed. Also added the new pkgs to ./hack/.linted_packages

sttts commented:
/lgtm

sttts commented on hack/.linted_packages:
```diff
@@ -493,3 +493,5 @@ test/integration/ttlcontroller
 test/soak/cauldron
 test/soak/serve_hostnames
 test/utils/junit
+staging/src/k8s.io/sample-apiserver/pkg/client/informers_generated/externalversions
```
can you sort this?

sttts commented on staging/src/k8s.io/sample-apiserver/hack/update-codegen.sh:
```diff
@@ -30,40 +35,50 @@ function cleanup {
+  local PREFIX=${BASE_PKG}/pkg/apis
+  local INPUT_BASE="--input-base ${PREFIX}"
+  local INPUT_APIS=(
+  ${GROUP_NAME}/
```
indent these two lines

sttts commented on staging/src/k8s.io/sample-apiserver/hack/update-codegen.sh:
```diff
@@ -30,40 +35,50 @@ function cleanup {
-${CLIENTGEN} ${INPUT_BASE} ${INPUT} ${CLIENTSET_PATH} 
-${CLIENTGEN} --clientset-name="clientset" ${INPUT_BASE} --input wardle/v1alpha1 ${CLIENTSET_PATH} 
+  local PREFIX=${BASE_PKG}/pkg/apis
+  local INPUT_BASE="--input-base ${PREFIX}"
```
don't like to have the flag name in the variable as well.

maleck13 replied:
Agree didn't like it either but it was the way it was written when I came to it, will update

sttts commented on staging/src/k8s.io/sample-apiserver/hack/update-codegen.sh:
```diff
@@ -18,10 +18,15 @@ set -o errexit
+informergen="${PWD}/informer-gen"
 # Register function to be called on EXIT to remove generated binary.
 function cleanup {
   rm -f "${CLIENTGEN:-}"
```
how about a temp directory for the binary? Then we get rid of these 3 ugly binary variables and we clean up listergen and informergen as well.

maleck13 replied:
Sure, but perhaps also in the follow up? I will create an issue with supporting multiple groups and versions and using a temp dir WDYT?

sttts replied:
sounds good

sttts commented:
/lgtm

Temp dir + multi group-version support in a follow-up.

maleck13 commented:
@sttts thanks for your reviews. I am trying to understand the failure in the verify job. It tells me 
20 BUILD files not up-to-date. Doing a bit of research perhaps I need to run ./hack/update-bazel.sh. Running that updates these files, should these vendor files also be committed?
Is there something I should read for making sure I know these things?

sttts commented:
`hack/update-bazel.sh` is right. You have to commit those files. They are für Google's build tool https://bazel.build/ that some Googlers use. I don't think those things are properly documented. In general for each verify step in CI there is an update script in hack/.

maleck13 commented:
Added the output from ```hack/update-bazel.sh``` also fixed the lint file as indicated by the failed build script

sttts commented:
@maleck13 can you create the follow-up issue?

maleck13 commented:
@sttts done

maleck13 commented:
Do all of these test normally pass? Not sure what I can do about the ```pull-kubernetes-e2e-kops-aws``` and ```pull-kubernetes-federation-e2e-gce```? Still trying to understand the issue with the bazel tests.

csbell commented:
/test pull-kubernetes-federation-e2e-gce

p0lyn0mial commented:
yeah, normally the tests are green. I think `pull-kubernetes-federation-e2e-gce` is flaky. 

https://github.com/kubernetes/kubernetes/pull/46874
https://github.com/kubernetes/kubernetes/issues/45954

p0lyn0mial commented on staging/src/k8s.io/sample-apiserver/pkg/client/clientset_generated/clientset/typed/wardle/v1alpha1/fischerlist.go:
```diff
@@ -0,0 +1,145 @@
+	Delete(name string, options *v1.DeleteOptions) error
+	DeleteCollection(options *v1.DeleteOptions, listOptions v1.ListOptions) error
+	Get(name string, options v1.GetOptions) (*v1alpha1.FischerList, error)
+	List(opts v1.ListOptions) (*v1alpha1.FischerListList, error)
```
seems like bazel does not like this line. from the logs `vendor/k8s.io/sample-apiserver/pkg/client/clientset_generated/clientset/typed/wardle/v1alpha1/fischerlist.go:41: undefined: v1alpha1.FischerListList`

cannot find corresponding line `FlunderListList`

sttts replied:
The annotation `// +genclient=true` above the `FischerList` type is wrong. It should only be above the non-list type.

maleck13 replied:
awesome thanks for the help

p0lyn0mial replied:
oh, of course that was so obvious :) - good catch!

out of curiosity, how it is possible that normal compilation via `make` works ?

sttts replied:
Only the generator uses the annotation. And as we throw away the clientset before this PR, everything is fine. Moreover, we don't use this clientset (yet) anywhere, so go build does not complain.

maleck13 commented:
will squash again once failing tests are passing

maleck13 commented:
Ok so all passing now except for the kops aws test. Anything I can do about that one?

fejta commented:
/test pull-kubernetes-e2e-kops-aws

maleck13 commented:
@sttts Looks like we are in good shape finally. Anything additional I need to do?

sttts commented:
/lgtm
